### DISCUSSION TITLE
Enhanced support to access MemberInfo of a Type

### DISCUSSION THREAD

**Discussion by GlobyteNet** [dotnet/csharplang#7269] [Language Ideas] [answered]
# Introduction

Currently in order to access MemberInfo of some type developers have to write this code: `typeof(MyType).GetProperty("MyProperty")` to access property and `typeof(MyType).GetMethod("MyMethod")` to access method.
This approach however has a few downsides:
1. No strong typing. If you misspelled the name, there is no hint. You can use `nameof` operator, however then the line of code becomes longer.
2. No hints of usage. Compiler doesn't recognize this property as used and if deleted, there will be no warning or error until you execute a code and see that it returns null which leads to third issue
3. GetProperty and GetMethod are defined as returning PropertyInfo? and MethodInfo?, which makes compiler complain about possible null dereferencing even if you are 100% sure that the member exists
With generic methods that are overloaded multiple times (like members of Enumerable, that you need to find to write dynamic linq queries) it becomes even more difficult to find them. For example, Mapster has this code:
``` csharp
//src.Select(item => convert(item))
var method = (from m in typeof(Enumerable).GetMethods()
where m.Name == nameof(Enumerable.Select)
let p = m.GetParameters()[1]
where p.ParameterType.GetGenericTypeDefinition() == typeof(Func<,>)
select m).First().MakeGenericMethod(sourceElementType, destinationElementType);
```
Without the provided comment it is difficult to understand, which exact method should be found. Also 5 lines of code to find a single method is a little bit too much.

## Improvement

The proposal is to provide simpler way to access MemberInfo by simply specifying strongly typed member signature. The idea is to extend typeof operator (another possible solution is to provide some new keyword). 
Now it will be possible to write `typeof(MyType.MyProperty)` which will return `PropertyInfo`. Similarly writing `typeof(MyType.MyMethod(int, string))` will return proper MethodInfo. Accessing generic methods could be done like that: `typeof(MyType.MyMethod<T>(int, T))`. Retireving bound MethodInfo would be possible if added specific types instead of placeholders: `typeof(MyType.MyMethod<int>(int, int))`. Another alternative syntax could be possible.

Providing this functionality will significantly reduce the size of code, when multiple MemberInfos are required, for example when generating Expression trees.

## Drawbacks
this could be a possible breaking change. To eliminate conflicts with existing code new syntax can be used like using colon
`typeof(MyType:MyMethod)`, which is currently not a valid code.

## Questions
How would be costructors, indexers and other members found? What would be the syntax for them?

**Comment by HaloFour** [comment, marked as answer]
See: #712

**Reply by GlobyteNet** [reply]
Thanks for the link. I knew I wasn't the first to come up with this idea and it's a shame to know that after so many years it still hasn't been implemented and probably never will be.
